src/v/iceberg/datatypes_json.cc

```diff
@@ -0,0 +1,251 @@
+    }
+    throw std::invalid_argument(fmt::format(
+      "Decimal requires format decimal(uint32, uint32): {}", type_str));
+}
+
+fixed_type parse_fixed(const std::string_view& type_str) {
+    auto l_str = extract_between('[', ']', type_str);
+    auto l = std::stoull(ss::sstring(l_str));
+    return fixed_type{l};
+}
+
+} // namespace
+
+struct_type parse_struct(const json::Value& v) {
+    struct_type ret;
+    const auto& fields_json = parse_required(v, "fields");
+    const auto& fields_array = fields_json.GetArray();
+    ret.fields.reserve(fields_array.Size());
+    for (const auto& field_json : fields_array) {
+        ret.fields.emplace_back(parse_field(field_json));
+    }
+    return ret;
+}
+
+list_type parse_list(const json::Value& v) {
+    const auto& element_json = parse_required(v, "element");
+    return list_type::create(
+      parse_required_i32(v, "element-id"),
+      parse_required_bool(v, "element-required") ? field_required::yes
+                                                 : field_required::no,
+      parse_type(element_json));
+}
+
+map_type parse_map(const json::Value& v) {
+    const auto& key_json = parse_required(v, "key");
+    const auto& val_json = parse_required(v, "value");
+    auto value_required = parse_required_bool(v, "value-required");
+    return map_type::create(
+      parse_required_i32(v, "key-id"),
+      parse_type(key_json),
+      parse_required_i32(v, "value-id"),
+      value_required ? field_required::yes : field_required::no,
+      parse_type(val_json));
+}
+
+nested_field_ptr parse_field(const json::Value& v) {
+    auto id = parse_required_i32(v, "id");
+    auto name = parse_required_str(v, "name");
+    auto required = parse_required_bool(v, "required");
+    const auto& type_json = parse_required(v, "type");
+    auto type = parse_type(type_json);
+    return nested_field::create(
+      id,
+      std::move(name),
+      required ? field_required::yes : field_required::no,
+      std::move(type));
+}
+
+field_type parse_type(const json::Value& v) {
+    if (v.IsString()) {
+        auto v_str = std::string_view{v.GetString(), v.GetStringLength()};
+        if (v_str.starts_with("decimal")) {
+            return parse_decimal(v_str);
+        }
+        if (v_str.starts_with("fixed")) {
+            return parse_fixed(v_str);
+        }
+        return string_switch<field_type>(v_str)
+          .match("boolean", boolean_type{})
+          .match("int", int_type{})
+          .match("long", long_type{})
+          .match("float", float_type{})
+          .match("double", double_type{})
+          .match("date", date_type{})
+          .match("time", time_type{})
+          .match("timestamp", timestamp_type{})
+          .match("timestamptz", timestamptz_type{})
+          .match("string", string_type{})
+          .match("uuid", uuid_type{})
+          .match("binary", binary_type{});
+    }
+    if (!v.IsObject()) {
```

Comment:
handle the default match case with an exception?

Reply:
This already appears to be handled? https://github.com/redpanda-data/redpanda/blob/015afa74f2d5dcf68b7c92b9dcabf28640edb5db/src/v/strings/tests/string_switch_test.cc#L58-L70

Reply:
ahh i thought it just fell through or something. sgtm